File: geo3x3.rs
use std::char::from_u32;

pub fn encode(lat: f64, lng: f64, level: i32) -> String {
	if level < 1 {
		return "".to_string();
	}
	let mut res = String::new();
	let mut lng2 = lng;
	if lng >= 0.0 {
		res.push_str("E");
	} else {
		res.push_str("W");
		lng2 += 180.0;
	}
	let mut lat2 = 90.0 - lat; // 0:the North Pole,  180:the South Pole
	let mut unit = 180.0;
	for _ in 1 .. level {
		unit /= 3.0;
		let x = (lng2 / unit) as i32;
		let y = (lat2 / unit) as i32;
		//println!("{}", from_u32((x + y * 3 + 1) as u32).unwrap());
		let s = from_u32(('0' as i32 + x + y * 3 + 1) as u32).unwrap();
		res.push(s);
		lng2 -= (x as f64) * unit;
		lat2 -= (y as f64) * unit;
	}
	return res.to_string();
}

pub fn decode(code: String) -> (f64, f64, i32, f64) {
	let mut unit = 180.0;
	let mut lat = 0.0;
	let mut lng = 0.0;
	let mut level = 1;

	let mut begin = 0;
	let mut flg = false;
	let c = code.chars().nth(0).unwrap();
	if c == '-' || c == 'W' {
		flg = true;
		begin = 1;
	} else if c == '+' || c == 'E' {
		begin = 1;
	}
	let clen = code.len();
	for i in begin .. clen {
		let c = code.chars().nth(i).unwrap();
		if c < '1' || c > '9' {
			break;
		}
		let n = c as i32 - '1' as i32;
		unit /= 3.0;
		lng += (n % 3) as f64 * unit;
		lat += (n / 3) as f64 * unit;
		level += 1;
	}
	lat += unit / 2.0;
	lng += unit / 2.0;
	lat = 90.0 - lat;
	if flg {
		lng -= 180.0;
	}
	return (lat, lng, level, unit);
}

File: geo3x3.zig
pub fn encode(lat: f64, lng: f64, level: u8, result: []u8) !void {
    if (level < 1) {
        return;
    }

    if (result.len < level) {
        return;
    }

    var _lng = lng;
    var _lat = lat;

    if (_lng >= 0.0) {
        result[0] = 'E';
    } else {
        result[0] = 'W';
        _lng += 180.0;
    }

    _lat += 90.0;
    var _unit: f64 = 180.0;

    var i: u8 = 1;
    while (i < level) {
        _unit /= 3.0;
        const x: u8 = @intFromFloat(_lng / _unit);
        const y: u8 = @intFromFloat(_lat / _unit);
        result[i] = '0' + x + y * 3 + 1;
        _lng -= @as(f64, @floatFromInt(x)) * _unit;
        _lat -= @as(f64, @floatFromInt(y)) * _unit;
        i += 1;
    }
}


pub fn decode(code: []const u8, result: []f64) !void {
    if (code.len == 0) {
        return;
    }

    var begin: u8 = 0;
    var flg: bool = false;

    const c = code[0];
    if (c == '-' or c == 'W') {
        flg = true;
        begin = 1;
    } else if (c == '+' or c == 'E') {
        begin = 1;
    }

    var unit: f64 = 180.0;
    var lat: f64 = 0.0;
    var lng: f64 = 0.0;
    var level: u8 = 1;

    var i: u64 = begin;
    while (i < code.len) {
        var n: u8 = code[i] - '0';
        if (n <= 0) break;
        unit /= 3.0;
        n = n - 1;
        lng += @as(f64, @floatFromInt(n % 3)) * unit;
        lat += @as(f64, @floatFromInt(n / 3)) * unit;
        level += 1;
        i += 1;
    }

    lat += unit / 2.0;
    lng += unit / 2.0;
    lat -= 90.0;
    if (flg) lng -= 180.0;

    result[0] = lat;
    result[1] = lng;
    result[2] = @as(f64, @floatFromInt(level));
    result[3] = unit;
}
